File: SooYeon/240101_week/Programmers_150369_택배 배달과 수거하기.js
function solution(cap, n, deliveries, pickups) {
    let answer = 0;
    
    // 배달과 수거 중 거리 값이 더 큰 값을 선택
    // 멀리있는 것 부터 확인
    while(deliveries.length || pickups.length) {
        let possible = cap;
        let dist = [0, 0];
        
        // 배달 거리 계산
        // 0인 인덱스는 없애버림
        for(let i = deliveries.length - 1; i >= 0; i--) {
            if(deliveries[i] !== 0) {
                // 거리 더하기
                // 가장 큰 인덱스(거리)로 저장
                if(possible === cap) dist[0] = i + 1;
                
                if(deliveries[i] <= possible) {
                    possible -= deliveries[i];
                    deliveries.pop();
                }
                else {
                    deliveries[i] -= possible;
                    break;
                }
            } 
            else deliveries.pop();
        }
        
        possible = cap; // 수거 계산을 위해 초기화
        
        // 수거 거리 계산
        // 0인 인덱스는 없애버림
        for(let i = pickups.length - 1; i >= 0; i--) {
            if(pickups[i] !== 0) {
                // 거리 더하기
                // 가장 큰 인덱스(거리)로 저장
                if(possible === cap) dist[1] = i + 1;
                
                if(pickups[i] <= possible) {
                    possible -= pickups[i];
                    pickups.pop();
                }
                else {
                    pickups[i] -= possible;
                    break;
                }
            }
            else pickups.pop();
        }
        
        // 배달과 수거 중 거리 값이 더 큰 값을 선택
        answer += Math.max(dist[0], dist[1]) * 2;
    }
    
    return answer;
}

File: SooBin/240101_week/programmers_150369_택배 배달과 수거하기.cpp
// 테스트 케이스 16,17 시간초과 못잡음

#include <string>
#include <vector>

using namespace std;

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    
    while (n > 0 && (deliveries[n - 1] == 0 && pickups[n - 1] == 0)) {
                n--;
    }
    
    while(n>0) {
        int delivery = cap;
        int pickup = cap;
        answer += 2 * n;
        
        for(int i = n-1; i >= 0; i--){
            
            if(deliveries[i] <= delivery){
                delivery -= deliveries[i];
                deliveries[i] = 0;
            } else {
                if(delivery > 0){
                    deliveries[i] -= delivery;
                    delivery = 0;
                }
            }
            if(pickups[i] <= pickup){
                pickup -= pickups[i];
                pickups[i] = 0;
            } else {
                if(pickup > 0){
                    pickups[i] -= pickup;
                    pickup = 0;
                }
            }
            
            while (n > 0 && (deliveries[n - 1] == 0 && pickups[n - 1] == 0)) {
                n--;
            }
            
        }
        
    }
  
    return answer;
}